q2_qemistree/_hierarchy.py

```diff
@@ -8,58 +8,87 @@
 
 import biom
 import pandas as pd
-from sklearn.metrics import pairwise_distances
-from scipy.spatial.distance import squareform
+from scipy.spatial.distance import squareform, pdist
 from scipy.cluster.hierarchy import linkage
 from skbio import TreeNode
+import numpy as np
 from q2_feature_table import merge
 
 from ._collate_fingerprint import collate_fingerprint
-from ._match import match_label
+from ._match import match_tables
 from ._semantics import CSIDirFmt
 
 
-def build_tree(relabeled_fingerprints: pd.DataFrame) -> TreeNode:
+def pdist_union(u, v):
+    b = np.double(np.bitwise_or(u, v).sum())
+
+    return b
+
+
+def pairwise_jaccard_modified(merged_fps: pd.DataFrame,
+                              merged_fdata: pd.DataFrame,
+                              mz_tolerance: float):
+    merged_fdata['row m/z'] = pd.to_numeric(merged_fdata['row m/z'])
+    merged_fdata = merged_fdata.sort_index()
+    merged_fps = merged_fps.sort_index()
+    jsim = 1 - pdist(X=merged_fps, metric='jaccard')
+    union = pdist(X=merged_fps != 0, metric=pdist_union)
+    mz_diff = pdist(X=merged_fdata['row m/z'].values.reshape(-1, 1),
+                    metric='cityblock')
+    mz_diff = (mz_diff <= mz_tolerance).astype(np.int32)
+    intersection = jsim * union
+    jmod = 1 - ((intersection+mz_diff)/(union+1))
+
+    return jmod
+
+
+def build_tree(merged_fps: pd.DataFrame,
+               merged_fdata: pd.DataFrame,
+               mz_tolerance: float) -> TreeNode:
     '''
     This function makes a tree of relatedness between mass-spectrometry
     features using molecular substructure fingerprints.
     '''
-    distmat = pairwise_distances(X=relabeled_fingerprints,
-                                 Y=None, metric='jaccard')
+
+    distmat = pairwise_jaccard_modified(merged_fps, merged_fdata, mz_tolerance)
     distsq = squareform(distmat, checks=False)
     linkage_matrix = linkage(distsq, method='average')
     tree = TreeNode.from_linkage_matrix(linkage_matrix,
-                                        relabeled_fingerprints.index.tolist())
+                                        merged_fps.index.tolist())
     return tree
 
 
-def merge_feature_data(fdata: pd.DataFrame):
+def merge_relabel(fps: pd.DataFrame, fts: pd.DataFrame, fdata: pd.DataFrame):
```

Comment:
Is it safe to assume that most users will have one single feature data object for all studies? Or should we assume that this is also a list of dataframes?